lib/src/command/add.dart

```diff
@@ -0,0 +1,339 @@
+      dependencies: dependencies,
+      devDependencies: devDependencies,
+      dependencyOverrides: original.dependencyOverrides.values,
+    );
+  }
+
+  /// Parse [package] to return the corresponding [PackageRange], as well as its
+  /// representation in `pubspec.yaml`.
+  ///
+  /// [package] must be written in the format
+  /// `<package-name>[:<version-constraint>]`, where quotations should be used
+  /// if necessary.
+  ///
+  /// Examples:
+  /// ```
+  /// retry
+  /// retry:2.0.0
+  /// retry:^2.0.0
+  /// retry:'>=2.0.0'
+  /// retry:'>2.0.0 <3.0.1'
+  /// 'retry:>2.0.0 <3.0.1'
+  /// retry:any
+  /// ```
+  ///
+  /// If a version constraint is provided when the `--path` or any of the
+  /// `--git-<option>` options are used, a [PackageParseError] will be thrown.
+  ///
+  /// Packages must either be a git, hosted, sdk, or path package. Mixing of
+  /// options is not allowed and will cause a [PackageParseError] to be thrown.
+  ///
+  /// If any of the other git options are defined when `--git-url` is not
+  /// defined, an error will be thrown.
+  Pair<PackageRange, dynamic> _parsePackage(String package) {
+    ArgumentError.checkNotNull(package, 'package');
+
+    if ([path != null, hasGitOptions, hasHostOptions, sdk != null]
+            .where((element) => element == true)
+            .length >=
+        2) {
+      usageException(
+          'Packages must either be a git, hosted, sdk, or path package.');
+    }
+
+    /// The package to be added, along with the user-defined package constraints
+    /// if present.
+    PackageRange packageRange;
+
+    /// The entry to be added to the pubspec. Assigned dynamic because it can
+    /// take on either a string for simple version constraints or a map for
+    /// more complicated hosted/git options.
+    dynamic pubspecInformation;
+
+    final splitPackage = package.split(':');
+    final packageName = splitPackage[0];
+
+    /// There shouldn't be more than one `:` in the package information
+    if (splitPackage.length > 2) {
+      throw FormatException('Invalid package and version constraint: $package');
+    }
+
+    /// We want to allow for [constraint] to take on a `null` value here to
+    /// preserve the fact that the user did not specify a constraint.
+    final constraint = splitPackage.length == 2
+        ? VersionConstraint.parse(splitPackage[1])
+        : null;
+
+    /// Determine the relevant [packageRange] and [pubspecInformation] depending
+    /// on the type of package.
+    if (hasGitOptions) {
+      dynamic git;
+
+      /// Process the git options to return the simplest representation to be
+      /// added to the pubspec.
+      if (gitRef == null && gitPath == null) {
+        git = gitUrl;
+      } else {
+        git = {'url': gitUrl, 'ref': gitRef, 'path': gitPath};
+        git.removeWhere((key, value) => value == null);
+      }
+
+      packageRange = cache.sources['git']
+          .parseRef(packageName, git)
```

Comment:
Can these `parseRef` functions throw exceptions? If so who is going to catch them...

What happens if we have a `pub add foo --git-ref` but no `--git-url` or `--git-url ""` or `--git-url !` or some other invalid value?

Reply:
hm the way we are setting things up, I don't think `parseRef` will throw an exception, because it checks mainly to see if the `git` variable is a String or `Map<String,String>` with the relevant keys, which will always be the case I think...?

Invalid values for `--git-url` are handled in the `acquireDependencies` function, the same way `pub get` does it at the moment.

I think what I'll do is to surround the `resolveVersions` call on the in-memory pubspec with a try catch, and on `GitException` I could print out a message asking the user to check the git arguments, and on `SolveFailure` (for path/hosted/sdk) I could print out a message asking the user to generally check the arguments - does this sound good?